opds2_import.py

```diff
@@ -0,0 +1,830 @@
+        rights_uri = RightsStatus.rights_uri_from_string("")
+        rel = first_or_default(link.rels, default_link_rel)
+        media_type = link.type
+        href = link.href
+
+        if feed_self_url and not urlparse(href).netloc:
+            # This link is relative, so we need to get the absolute url
+            href = urljoin(feed_self_url, href)
+
+        link_metadata = LinkData(
+            rel=rel,
+            href=href,
+            media_type=media_type,
+            rights_uri=rights_uri,
+            content=None,
+        )
+
+        self._logger.debug(
+            u"Finished extracting link metadata from {0}: {1}".format(
+                encode(link), encode(link_metadata)
+            )
+        )
+
+        return link_metadata
+
+    def _extract_description_link(self, publication):
+        self._logger.debug(
+            u"Started extracting a description link from {0}".format(
+                encode(publication.metadata.description)
+            )
+        )
+
+        description_link = None
+
+        if publication.metadata.description:
+            description_link = LinkData(
+                rel=Hyperlink.DESCRIPTION,
+                media_type=MediaTypes.TEXT_PLAIN,
+                content=publication.metadata.description,
+            )
+
+        self._logger.debug(
+            u"Finished extracting a description link from {0}: {1}".format(
+                encode(publication.metadata.description), encode(description_link)
+            )
+        )
+
+        return description_link
+
+    def _extract_image_links(self, publication, feed_self_url):
+        """Extracts a list of LinkData objects containing information about artwork.
+
+        :param publication: Publication object
+        :type publication: ast_core.Publication
+
+        :param feed_self_url: Feed's self URL
+        :type feed_self_url: str
+
+        :return: List of links metadata
+        :rtype: List[LinkData]
+        """
+        self._logger.debug(
+            u"Started extracting image links from {0}".format(
+                encode(publication.images)
+            )
+        )
+
+        sorted_raw_image_links = list(
+            reversed(
+                sorted(
+                    publication.images.links, key=lambda link: (link.width, link.height)
+                )
+            )
+        )
+        image_links = []
+
+        if len(sorted_raw_image_links) > 0:
+            cover_link = self._extract_link(
+                sorted_raw_image_links[0],
+                feed_self_url,
+                default_link_rel=Hyperlink.IMAGE,
+            )
+            image_links.append(cover_link)
+
+        if len(sorted_raw_image_links) > 1:
+            cover_link = self._extract_link(
+                sorted_raw_image_links[1],
+                feed_self_url,
+                default_link_rel=Hyperlink.THUMBNAIL_IMAGE,
+            )
+            image_links.append(cover_link)
```

Comment:
I see what you're doing -- pick the biggest image to be the main one and a smaller version to be the thumbnail -- but I don't think this will work in general. There's no guarantee that these images have the same media type, or that the second-largest image isn't far too large to use as a thumbnail.

At the same time I don't know of a better solution. Ideally the CM would import the biggest image and make its own thumbnail, but there's no guarantee that the CM has any way of mirroring thumbnails, since that's an optional feature.

We have a few constants for the ideal size and aspect ratio for a thumbnail image:

```
    IDEAL_COVER_ASPECT_RATIO = 2.0/3
    IDEAL_IMAGE_HEIGHT = 240
    IDEAL_IMAGE_WIDTH = 160
```

So put in a TODO to fix this up using those constants. Instead of using the second-largest image as a thumbnail, find the image that would make the best thumbnail because of its dimensions and its media type.

Reply:
Added the comment